fix timout by using our own input buffer

khera opened the pull request:
The timeout was improperly implemented using select() on a buffered file
handle. This cannot work because there will be data to use in the buffer even
if the handle is not ready to read, thus we deadlock.

We must use our own buffering so we know to call select() only when we are
certain we need to fetch more data from the handle.

khera committed:
fix timout by using our own input buffer

The timeout was improperly implemented using select() on a buffered file
handle. This cannot work because there will be data to use in the buffer even
if the handle is not ready to read, thus we deadlock.

We must use our own buffering so we know to call select() only when we are
certain we need to fetch more data from the handle.

carwynmoore commented on lib/Net/NATS/Connection.pm:
```diff
@@ -35,6 +45,125 @@ sub upgrade {
+    if (!defined $count) {
+      return '0E0' if $! == EWOULDBLOCK; # we handle this error
+
+      $self->error($!);                  # remember the error for later
```
$self->error is an lvalue, so this line needs to be:

```
$self->error = $!;
```

carwynmoore commented:
Am seeing some errors when publishing, either truncated data, or too much data.  Have included the log output from gnats which includes the last successful publish, and then the publish that causes a parsing error.

```
[95952] 2016/07/08 17:57:00.539477 [TRC] 127.0.0.1:35663 - cid:1 - ->> [PUB foo 13]
[95952] 2016/07/08 17:57:00.539493 [TRC] 127.0.0.1:35663 - cid:1 - ->> MSG_PAYLOAD: [Hello, World!]
PUB foo 13]6/07/08 17:57:00.539536 [TRC] 127.0.0.1:35663 - cid:1 - ->> [PUB foo 13
PUB foo 13'6/07/08 17:57:00.539594 [ERR] 127.0.0.1:35663 - cid:1 - Error reading from client: processPub Parse Error: 'foo 13
```

```
[95952] 2016/07/08 17:58:07.003507 [TRC] 127.0.0.1:35749 - cid:5 - ->> [PUB foo 13]
[95952] 2016/07/08 17:58:07.003513 [TRC] 127.0.0.1:35749 - cid:5 - ->> MSG_PAYLOAD: [Hello, World!]
[95952] 2016/07/08 17:58:07.003519 [TRC] 127.0.0.1:35749 - cid:5 - ->> [PUB foo 13]
[95952] 2016/07/08 17:58:07.004000 [ERR] 127.0.0.1:35749 - cid:5 - Error reading from client: Client Parser ERROR, state=30, i=23199: proto='"\nHello, World!\r\nPUB foo 13\r\nHell"...'
```

khera commented:
I did not alter the publish (output) side at all. All I updated was the
reading (input) routines. I just replaced the `read_line()` and `read()`
methods from the client, and implemented them using primitives I added to
the connection object.

Could you share which test script you were running that caused this? I'll
try to reproduce.

khera committed:
incorrectly using error lvalue

During development, I changed the error accessor to be an lvalue, but
forgot to update where it was set in two places. This is fixed.

khera commented:
I found a problem with the writes... because the socket is marked as non-blocking, syswrite will occasionally return "EAGAIN" when it cannot complete the write. I'm working up a solution to this. I'm trying to do it without dispatching to another method as that seems to incur an almost 21% penalty in the speed of publish().

khera committed:
Handle non-blocking write

The I/O handle is marked as non-blocking, so syswrite will on occasion return
EAGAIN. This is now handled. I also handle EINTR since that's always possible
and was cheap to test.

This code as committed is correct, however I'm not 100% satisified with the
structure of having some I/O in NATS::Client and some in NATS::Connection.

khera committed:
restructure send code
moved the send() method into the Connection object so that all I/O is in one place.

khera commented:
I fixed the bug with send. I moved around a little bit of code so all I/O is done inside the Connection object. I'm not totally sure how you like to structure the code, but this made sense to me.

Overall, the change to the publish() method has made it significantly slower according to the benchmark-pub.pl script you include. With the blocking write, I can get almost 700k publish/s but with this "safe" code it drops to around 490k/s.

The code as written now is correct and very solid based on my testing. I cannot make it lose messages outside the guarantees of NATS itself.

carwynmoore commented:
@khera Change is looking solid as, nice 👍 

The performance change can be looked at later, having your timeout change is more important at the moment.
